src/protocols/local.js

```diff
@@ -0,0 +1,224 @@
+ * Local Authentication Protocol
+ *
+ * The most widely used way for websites to authenticate users is via a username
+ * and/or email as well as a password. This module provides functions both for
+ * registering entirely new users, assigning passwords to already registered
+ * users and validating login requesting.
+ *
+ * For more information on local authentication in Passport.js, check out:
+ * http://passportjs.org/guide/username-password/
+ */
+
+var EMAIL_REGEX =
+  /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i
+
+/**
+ * Use validator module isEmail function
+ *
+ * @see <https://github.com/chriso/validator.js/blob/3.18.0/validator.js#L38>
+ * @see <https://github.com/chriso/validator.js/blob/3.18.0/validator.js#L141-L143>
+ */
+function validateEmail(str) {
+  return EMAIL_REGEX.test(str)
+}
+
+function generateToken() {
+  return base64URL(crypto.randomBytes(48))
+}
+
+/**
+ * Register a passport
+ *
+ * This method creates a new passport from a specified email, username and password
+ * and assign it to a newly created user.
+ *
+ */
+const createPassport = async function (user, password, accessToken) {
+  let result = {error: null, user: null}
+  return await PassportModelAPI.create({
+    protocol: 'local',
+    password: password,
+    user: user.id,
+    accessToken: accessToken
+  })
+    .then(async function (passport) {
+      result.user = user
+      return result
+    })
+    .catch(err => {
+      result.error = err
+      return result
+    })
+}
+
+/**
+ * Update a passport
+ *
+ * This method updates a passport of a specific user
+ *
+ */
+const updatePassport = async function (user, passport) {
+  let result = {error: null, user: null}
+  return PassportModelAPI.updateOne(passport)
+    .then(function (passport) {
+      result.user = user
+      return result
+    })
+    .catch(err => {
+      result.error = err
+      return result
+    })
+}
+
+/**
+ * Register a new user
+ *
+ * This method creates a new user from a specified email, username and password
+ * and assign the newly created user a local Passport.
+ *
+ */
+export const createUser = async function (_user) {
```

Comment:
Shouldn't we place this in the user model file ?

Reply:
cool done :)